core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -105,149 +105,72 @@ sealed trait Channel[F[_], A] {
-                    (
-                      State(a :: values, size + 1, None, producers, false),
-                      notifyStream(waiting)
-                    )
-                  else
-                    (
-                      State(values, size, None, (a, producer) :: producers, false),
-                      notifyStream(waiting) <* waitOnBound(producer, poll)
-                    )
-              }.flatten
-            }
+        // doesn't interrupt taking in progress
+        def close: F[Either[Channel.Closed, Unit]] =
+          closedR.complete(()).map {
+            case false => Left(Channel.Closed)
+            case true  => Right(())
           }
 
-        def close =
-          state
-            .modify {
-              case s @ State(_, _, _, _, closed @ true) =>
-                (s, Channel.closed.pure[F])
-
-              case State(values, size, waiting, producers, closed @ false) =>
-                (
-                  State(values, size, None, producers, true),
-                  notifyStream(waiting) <* signalClosure
-                )
-            }
-            .flatten
-            .uncancelable
-
-        def isClosed = closedGate.tryGet.map(_.isDefined)
-
-        def closed = closedGate.get
-
-        def stream = consumeLoop.stream
-
-        def consumeLoop: Pull[F, A, Unit] =
-          Pull.eval {
-            F.deferred[Unit].flatMap { waiting =>
-              state
-                .modify { state =>
-                  if (shouldEmit(state)) (empty(state.closed), state)
-                  else (state.copy(waiting = waiting.some), state)
-                }
-                .flatMap {
-                  case s @ State(values, stateSize, ignorePreviousWaiting @ _, producers, closed) =>
-                    if (shouldEmit(s)) {
-                      var size = stateSize
-                      var allValues = values
-                      var unblock = F.unit
-
-                      producers.foreach { case (value, producer) =>
-                        size += 1
-                        allValues = value :: allValues
-                        unblock = unblock <* producer.complete(())
+        def isClosed: F[Boolean] = closedR.tryGet.map(_.isDefined)
+
+        def send(a: A): F[Either[Channel.Closed, Unit]] =
+          isClosed.ifM(Channel.Closed.asLeft[Unit].pure[F], q.offer(a).map(Right(_)))
+
+        def stream: Stream[F, A] = {
+          val takeN: F[Chunk[A]] =
+            q.tryTakeN(None).flatMap {
+              case None =>
+                val fallback = MonadCancel[F].uncancelable { poll =>
+                  poll(Spawn[F].racePair(q.take, closedR.get)).flatMap {
+                    case Left((oca, fiber)) =>
+                      oca.embedNever.flatMap(a => fiber.cancel.as(Chunk.singleton(a)))
+
+                    case Right((fiber, ocb)) =>
+                      ocb.embedNever.flatMap { _ =>
+                        (fiber.cancel *> fiber.join).flatMap { oca =>
```

Comment:
I think this works if `fiber.cancel` arrives after `take` has successfully produced the elements, in which case it will be a no-op, and then the subsequent `join` gets you the elements. However imagine this:
`cancel` hits after the elements have arrived, but before `take` can complete. Due to the take invariant, the elements aren't lost and go back to the queue, but this code is still buggy: we will emit an empty chunk from here, and terminate the stream.

 We can try and do an additional `tryTake` in that case, but I haven't thought if there are further corner cases there. I think the take away here is that separating the state makes things very very complex to reason about, so before including this I'd want to see extremely convincing performance results :)

Reply:
> Due to the take invariant, the elements aren't lost and go back to the queue, but this code is still buggy: we will emit an empty chunk from here, and terminate the stream.

Ah yes, I actually thought about this but didn't implement the fix. We basically need a robust bit of signalling than the size of the `Chunk`. When we hit this case, we need to basically retry the `tryTake`.

> so before including this I'd want to see extremely convincing performance results :)

Agreed, though there's also value in reducing the maintenance burden and getting the benefit of *future* improvements in `Queue`.

> Btw, I'm not a hundred percent sure that invariant actually holds in the current version of queue, there's a race between deferred.complete and cancel a deferred.get, and get being cancelled cannot communicate to complete to put the elements back...

That's a bug then and we should fix it. We *really* need that invariant in general, and not just in this use of the queue.

Reply:
Yeah I'd rather merge if possible, it's just that this implementation strategy seems both unavoidable and risky, as its based essentially on recovering from interruption, and that's how I originally started with groupWithin and then had to change


As per the take guarantee, I'm not sure it's achievable without a locking implementation of queue, think about the two ops on Deferred racing _after_ they both have relinquished control of the queue state. Maybe it can be done with a more complex state machine in the Ref to encode the extra communication, but then how easily can that be ported to the low level implementation of queue?

Reply:
I *think* that the invariant should be achievable if we correctly manage our `uncancelable` regions. I'm not sure if we're doing that right now though (in fact, I'm certain that in some cases, we're *not* doing that in the `Async` queue). Need to stare at it a bit more. It feels like `Queue` itself would be fundamentally leaky if we can't achieve that invariant, though.

Reply:
Digging in a little deeper still, I don't think that `async` *itself* preserves this property. In fact, there's a specific place in the state machine where it catches the race condition between cancelation and completion... and resolves it in favor of cancelation. I need to think about this more, actually, because I honestly wonder if we're in a place where it's impossible to avoid data loss in all cases.

Reply:
Spamming this single thread... I think we can do this but it has to be encoded more like what the `Async` queue and the old `Channel` implementation does. Basically, the `Deferred` *must* contain a `Unit` because otherwise it simply isn't guaranteed to not lose data. Alternatively, we can probably fix `Deferred` itself.

Reply:
The Deferred of Unit is where my thoughts were going as well before I had to step afk

Reply:
>  If the taker.get is canceled after it has been fulfilled

This is not the problem though. The enqueue method will spin on a Ref whose state contains an unfulfilled Deferred, select it for delivery, and commit on the Ref. At this point, _Queue is no longer at play_. You now have a race between `deferred.complete` and cancelling a `deferred.get`, and it's perfectly possible that cancelling will win, i.e. it `get` gets cancelled _before_ being fulfilled, at which point `complete` does fulfill, but the element has been out of the queue for while, and can be put back in A) because `complete` doesn't know that `get` has been canceled, and `B` because even if it knew the order of elements in the queue might have changed (i..e another element might have been delivered) so you violate FIFO: you'd need to lock the Queue until you know the element has been successfully delivered, that's what I mean by wanting at least once processing.

An alternative would be to alter `take` instead to only allow cancellation when the relative Deferred is still in the state. It gets tricky cause you can't just return `Deferred.get`, you have to start dancing with extra fibers so that when the cancelation signal hits you first check that the Deferred is indeed still in the queue and hasn't been removed by enqueue about to complete it. But even then I think there is a sliver where after you've checked, found it the queue, then cancel, enqueue will complete, resulting in data loss again

Reply:
Need to reread and digest more, but I'm thinking of something a little simpler to get at this invariant. In particular, being very careful about where cancelation becomes visible. Basically, from the moment the element is committed to no longer being in the queue all the way to the end of `take`, cancelation must not be observed. If you preserve this property, then it's possible to have the "all or nothing" semantic in a useful way due to adjacency.

This seems actually doable. Hard but doable.

Reply:
That's what the second paragraph is exploring, but the issue is the element is no longer in the queue as part of a `Ref.modify`, where you can't execute effects, and so you cannot have an effect transactionally with the change in state, whereas you need to mark the Deferred as now uncancelable transactionally with the removal of an element in the queue

Reply:
I think that's only true though if we _don't_ restructure the Concurrent Queue to use `Deferred[Unit]`. The fact that there are multiple paths for values is, I think, the root of this issue.

Reply:
> I think that's only true though if we don't restructure the Concurrent Queue to use Deferred[Unit]. The fact that there are multiple paths for values is, I think, the root of this issue.

I agree, but `Deferred[Unit]` brings its own challenges, that I alluded to in:

>  he Deferred of Unit is where my thoughts were going as well before I had to step afk, so you have a consistent place for all the state, but it's not trivial to do in the multiple consumers case,you have to mark elements as "takeable" for a given consumer to preserve FIFO.

In particular consider the multiple consumers case, and the preservation of FIFO: consumer 1 takes on empty queue, deferred gets enqueued, consumer 2 takes on empty queue, deferred gets enqueued. At this point, two elements are enqueued immediately after one another, which means the two Deferred are completed, and the consumers awake, at which point they _race_ on the Ref to take the element, which means that they can now break FIFO. You can think about communicating to each consumer the position of  the element of the queue they have to take, but this can be broken because you can communicate position 1 to consumer 1, position 2 to consumer 2, then consumer 1 gets cancelled, at which point consumer 2 gets position 2, whereas it should have switched to taking position 1. The reason the pattern works so well in the current Channel is because there is a single consumer, so "there are changes" is enough signal for it to work, but it gets very tricky in Queue

Reply:
IMO whenever you have multiple consumers racing for elements, FIFO goes out the window just a bit. It also becomes a lot less relevant because of the racing, so the lack of strong ordering across consumers is effectively unobservable.

Reply:
This is not the case here though, look at the example again, with Deferred of Unit two consumers who subscribed at wildly different times (say three seconds apart) end up racing when *production* is fast because after they wake up they have to spin on the Ref again, which would in a very observable bug

Reply:
I get what you mean; I was being imprecise about what I was referencing. Any number of suspended consumers, regardless of when they subscribe, race *at publication time* on data. Notably, this isn't a fairness issue, since suspended consumers will still be awakened in FIFO order, but it does mean that the multiplexing might be in a different order.

You're basically describing a situation where data **A** and **B** came into the queue *from two publishers* almost simultaneously, but with **A** still arriving first. Meanwhile, you have consumers **1** and **2** who subscribed in that order, and consumer **1** might get **B** instead of **A**. I really don't see why this would cause a serious or even noticeable bug, particularly since other mpmc implementations have analogous issues (e.g. the jctools queue has this same problem). From a parallelism perspective, there really isn't a strong ordering between **A** and **B** since they are arriving on separate publishers so closely interspersed, and thus which consumer gets which one is somewhat nondeterministic.

I would be more concerned if **A** and **B** could arrive on the *same* publisher and the ordering could be mixed up on the consumers, but I don't believe that is possible.

Reply:
> I would be more concerned if A and B could arrive on the same publisher and the ordering could be mixed up on the consumers, but I don't believe that is possible.

That's exactly what I'm saying will happen with `Deferred[F, Unit]` :)

With `Deferred[F, Unit]` , a publisher wakes a consumer up, then the consumer has to spin again on the Ref to get the value. Crucially, `deferred.complete` terminates well before this happens: `complete` terminates after calling the callback, at which point `get` will awake, and recur back into a `modify`on `Ref`.

Things have this general shape:

```scala
enqueue = transactionally { 
    put elem in elem queue
    take first Deferred[Unit] from listeners queue
} then complete Deferred[Unit]

get = transactionally {
  full => take elem
  empty => enqueue Deferred[Unit]
  } then wait on Deferred[Unit], and recur on get
```

So let's see what happens: Consumer 1 takes on the empty queue, and enqueues a deferred, Consumer 2 (even after some time) takes on the empty queue, and enqueues a deferred, and after some time a single Producer produces two elements, let's break that down:
1) first transaction on enqueue, Consumer 1's deferred gets dequeued
2) Consumer 1 deferred gets completed, Producer can keep going
... At this point Consumer 1's recurring on the Ref is racing with  Producer enqueing second element, imagine Producer wins...
3) Producer's second transaction on enqueue, Consumer 2's deferred gets dequeued
4) Consumer's 2 deferred gets completed,
... now Consumer's 2 recurs on the Ref

At this point both Consumer 1 and Consumer 2 are racing to `modify` on the Ref, and if Consumer 2 wins, FIFO will be broken, even though Consumer 1 and Consumer 2 called dequeue at sufficiently distant times, and even though there is a single producer operating strictly synchronously, which to me sounds exactly like a race condition bug in the implementation.

What you seem to want is `deferred.complete` effectively blocking until the element it's signalling is successfully dequeued, which once again points to a blocking implementation.

Reply:
Also note that the problem I'm describing does not happen in the current design, because `Deferred[F, A]` means that you transactionally decide which element goes where, so FIFO is respected: by the time the first produce transaction finishes and the second produce transaction starts, the element is already going to the correct destination. There is an unavoidable (I think) tradeoff here: the `Deferred[F, A]` does FIFO correctly because the elements is considered delivered in `modify`, and by the same token if things get cancelled later, the element is "lost" (because it's already considered delivered, there is no extra acking). `Deferred[F, Unit]`  is more lenient with cancellation because the element isn't considered delivered after produce's `modify`, but that's also the reason why it breaks FIFO. I think the current queue implementation has the right tradeoff, and what's unreasonable is expecting to be able to race a `take` and retry without losing elements, by wishing more transactionality on it than possible, which again is trying to cover the fact that one lost transactionality by splitting the concurrent state in two (the `closed` signal, and the `queue`)

Reply:
Ah I see what you're getting at.

Okay, real talk: how much would we really care if that were the case? Because not being able to implement something like Channel on top of Queue feels much worse. Losing data without any opportunity to do it "right" definitely feels worse.

Reply:
> Okay, real talk: how much would we really care if that were the case? 

I mean, we're talking about breaking a primary characteristic of what defines a queue, in a case with no contention. I bet if someone reported that right now, you'd treat it as a bug

> Because not being able to implement something like Channel on top of Queue feels much worse.

Well, I can agree that not having STM without its performance cost is not as nice as having it (because that's what we're talking about here), but it's also the sad reality of things. We could try a hybrid implementation strategy that uses both a closed flag and wraps elems in `Option` and see if that makes things behave more reasonably without imposing a perf cost, but that's hard to say because we don't have have convincing perf numbers yet, although perf numbers with relaxed guarantees are also hard to discuss 

> Losing data without any opportunity to do it "right" definitely feels worse.

This is the bit where, regardless of the decision for Channel, I think the expectation is unrealistic. I see it as the same as the operation processing the element of `take` failing immediately: i.e. `Queue` has at most once semantics, and really the implementation strategy of `take-interrupt-tryTake` wants at least once